src/main/java/org/opensearch/neuralsearch/query/ModelInferenceQueryBuilder.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+/**
+ * Query builders which calls ml-commons API to do model inference.
+ * The model inference result is used for search on target field.
+ */
+
+public interface ModelInferenceQueryBuilder {
```

Comment:
Makes sense. @navneet1v @jmazanec15 @zane-neo Do you have any preferences or concerns for these two implementation?

Reply:
If we create a new abstract class, we can extract the common attributes of fieldName, queryText, and modelId. And we can also wrap the serialize/deserialize of these fields. It seems more close to the mental model of these 2 query clauses.

The interface way requires less effort, and it has smaller risk in implementation. 

I'm open to both

Reply:
Agree with the first option since the interface isn't add any concrete functionality to the class, in this case we shouldn't use interface and all the methods are retrieving attributes of a class(Builder), so we should have a class to hold these attributes and make it reusable.

Reply:
We do have mechanism for handling default model id via visitor in the QueryBuilder. Please check [this PR](https://github.com/opensearch-project/neural-search/pull/541) where support for default model id has been added to hybrid query. Can we use same approach here, or if not - please explain why.

Reply:
This PR support default model id of neural query nested inside hybrid query. The prerequisite is neural query default model id has already been supported. Current code check the class and support NeuralQueryBuilder only

Reply:
https://github.com/opensearch-project/neural-search/blob/ea49d3c5006efff9dfa36e69791ae9a8e468d25a/src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java#L31

Reply:
Hey @zhichao-aws I think we do not need this interface. Neither we need an abstract class. I would suggest you to just create a new visitor.

Just like NeuralSearchQueryVisitor
Create one NeuralSparseSearchQueryVisitor.

NeuralQueryBuilder and NeuralSparseQueryBuilder has different purposes. In future, it might happen parameters change  in either of them. So it should be kept independent.

We have throughly tested NeuralSearchQueryVisitor. Making changes in that design might affect functionality.

So it is better if we create a NeuralSparseQueryVisitor. 

Also there are certain things to note

Every visitor has a getChildVisitor method which has Occur occur as a argument.

Tomorrow just for NeuralQueryBuilder if cx asks for a certain thing to happen at Occur.Must then we can restrict that change to just NeuralQuery. This will not impact NeuralSparseQueryBuilder. Therefore, keeping them independent is for greater good.


cc: @martin-gaievski @navneet1v

Reply:
Hi @vibrantvarun , I have some concerns to implement a dedicated new visitor will share all codes with the current one.
1. For current implementation, a dedicated new visitor will share all codes with the current one, except the class checker. This will make our code base redundant. And bring more maintaining efforts. If we want to introduce new feature or fix bugs, we need to add code in both places.
2. Although NeuralQueryBuilder and NeuralSparseQueryBuilder has different purposes, they have much in common. Using an interface to extract the shared part is consistent with the mental model. The parameters may change for each query builder, but I don't think we'll change the most basic fields (in this PR, modelId and fieldName). Even if they are changed, we can still add more logics here to distinguish these two query builders.
For the risk point, I think the interface way is safe enough. We can regard it as a marker of these 2 classes, this will minimize the impact to existing logics. We have many test cases to ensure the functionality of existing features (I can also add new tests if necessary).

Reply:
Please add BWC tests.

Reply:
+1